File: RightslineDemoAppDotNet/RightslineDemoAppDotNetSQS/RestClient.cs
using RightslineDemoAppDotNetSQS.Config;
using System;
using System.Collections.Generic;
using System.Dynamic;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Runtime.CompilerServices;
using System.Text;
using System.Threading.Tasks;
using AWSSignatureV4_S3_Sample.Signers;
using AWSSignatureV4_S3_Sample.Util;
using Microsoft.VisualBasic.CompilerServices;
using System.Timers;
using Newtonsoft.Json;
using System.Xml;
using Newtonsoft.Json.Linq;
using System.Text.RegularExpressions;
using System.Net;
//
// SOURCE : https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-examples-using-sdks.html
//
namespace RightslineDemoAppDotNetSQS
{
    //For this demo app, we will not be using the Amazon AWS SDK
    //We recommend using the SDK but we have created this example in case you are restricted from using it
    public class RestClient
    {
        private static string BaseUrl = "https://sqs.us-west-2.amazonaws.com/";
        private static Dictionary<string, string> config = ConfigSetup.GetConfigFile();
        private static int SecondsToPoll = 5;
        private static string Regex = "\\{(.*?)\\}<";
        private static int MessagesToReceieve = 10;

        //This returns an XML string from the SQS queue that contains the 10 most recent messages. 
        public static string GetSQSMessages()
        {
            string region = config["Region"];
            // Construct a virtual hosted style address with the bucket name part of the host address,
            // placing the region into the url if we're not using us-east-1.
            var regionUrlPart = string.Format("-{0}", region);
            //This is our QA fifo queue, replace it with your queue's url
            var endpointUri = BaseUrl + config["AccountId"] + "/" + config["QueueName"];
            var requestParameters = "Action=ReceiveMessage&MaxNumberOfMessages=" + MessagesToReceieve;
            var uri = new Uri(endpointUri + "?" + requestParameters);            
            var headers = new Dictionary<string, string>()
            {
                {AWS4SignerBase.X_Amz_Content_SHA256, AWS4SignerBase.EMPTY_BODY_SHA256}
//                {"content-type", "text/plain"}
            };
            //Use the AWS4 signer to generate the signer and sign the request
            var signer = new AWS4SignerForAuthorizationHeader()
            {
                EndpointUri = uri,
                HttpMethod = "GET",
                Service = "sqs",
                Region = region
            };

            var authorization = signer.ComputeSignature(headers, requestParameters, AWS4SignerBase.EMPTY_BODY_SHA256,
                config["AccessKey"], config["SecretKey"]);
            headers.Add("Authorization", authorization);

            string response = HttpHelpers.InvokeHttpRequest(uri, "GET", headers, null);
            //Console.WriteLine(response);
            return response;
        }

        //We start a thread to poll the FIFO queue every few seconds for new messsages 
        [STAThread]
        public static void StartBackgroundMonitoring()
        {
            //This is a timer that will poll the SQS queue every X seconds
            Timer t = new Timer(SecondsToPoll * 1000); // 1 sec = 1000, 60 sec = 60000
            t.AutoReset = true;
            t.Elapsed += new System.Timers.ElapsedEventHandler(t_Elapsed);
            t.Start();
        }
        // Valid entities
        // table, rightset, catalog-item, relationship, deal
        private static void t_Elapsed(object sender, System.Timers.ElapsedEventArgs e)
        {
            string entity = "catalog-item";
            var messages = GetSQSMessages();
            var jsonobjects = FilterXML(messages);
            Notify(jsonobjects, entity);
            var receipts = GetReceiptHandles(messages);
            DeleteMessage(receipts);
        }
        //Test method to check the queue once for a relationship item and delete received messages
        public static void DemoMonitor()
        {
            var messages = GetSQSMessages();
            var jsonobjects = FilterXML(messages);
            Notify(jsonobjects, "relationship");
            var receipts = GetReceiptHandles(messages);
            DeleteMessage(receipts);
        }
        //Uses regex to filter the XML then converts it to JSON and builds JSON objects out of the results
        private static List<JObject> FilterXML(string messages)
        {
            Regex filterRegex = new Regex(Regex);
            var filteredMessages = filterRegex.Matches(messages);
            var messageValues = new List<string>();
            foreach (Match message in filteredMessages)
            {
                messageValues.Add(message.Value.Substring(0, message.Length - 1));
            }
            var jsonObjects = new List<JObject>();
            foreach (string message in messageValues)
            {
                jsonObjects.Add(JObject.Parse(message));
            }
            return jsonObjects;
        }
        //Filters the XML message to retrieve receipt handles
        private static List<string> GetReceiptHandles(string messages)
        {
            Regex filterRegex = new Regex("ReceiptHandle>(.*?)<\\/Receipt");
            var receiptHandles = filterRegex.Matches(messages);
            var receipts = new List<string>();
            foreach (Match message in receiptHandles)
            {
                receipts.Add(message.Groups[1].ToString());
            }
            return receipts;
        }
        /// <summary>
        /// Writes a message to the console whenever a message in the queue regarding a given an entity shows up
        /// </summary>
        /// <param name="entityName">Name of the entity to be searching for</param>
        private static void Notify(List<JObject> messages, string entityName)
        {
            Console.WriteLine("Poll at " + DateTime.Now.ToString());
            int numMessages = 0;
            var entityregex = new Regex("v2\\/(.*?)\\/");
            foreach (JObject message in messages)
            {
                var x = entityregex.Match(message["entityUrl"].ToString());
                //message["entityUrl"].ToString().Contains(entityName)
                var messageEntity = entityregex.Match(message["entityUrl"].ToString()).Groups[1];
                if (messageEntity.ToString().Equals(entityName))
                {
                    numMessages++;
                    Console.WriteLine("A " + messageEntity + " was " + message["action"] + ", URL: " + message["entityUrl"]);
                }
                //Uncomment this if you want to be notified of all other messsages in the response
                //else
                //{
                //    Console.WriteLine(message["action"] + "  " + message["entityUrl"]);
                //}                
            }
            if (numMessages == 0)
            {
                Console.WriteLine("Recieved " + messages.Count + " messages. No messages regarding " + entityName + " entities were recieved");
            }
        }
        //Starts a thread to send a message to the FIFO queue to delete messages that were receieved
        [STAThread]
        private static void DeleteMessage(List<string> receiptHandles)
        {
            foreach (string receipt in receiptHandles)
            {
                string region = config["Region"];
                // Construct a virtual hosted style address with the bucket name part of the host address,
                // placing the region into the url if we're not using us-east-1.
                var regionUrlPart = string.Format("-{0}", region);
                //This is our QA fifo queue, replace it with your queue's url
                var endpointUri = BaseUrl + config["AccountId"] + "/" + config["QueueName"];
                //Encode the receipt handle and add it to the url
                Console.WriteLine(receipt);
                var encodedReceipt = WebUtility.UrlEncode(receipt);
                var requestParameters = "Action=DeleteMessage&ReceiptHandle=" + encodedReceipt;
                Console.WriteLine(encodedReceipt);
                var uri = new Uri(endpointUri + "?" + requestParameters);
                var headers = new Dictionary<string, string>()
                {
                    {AWS4SignerBase.X_Amz_Content_SHA256, AWS4SignerBase.EMPTY_BODY_SHA256}
//                  {"content-type", "text/plain"}
                };
                //Use the AWS4 signer to generate the signer and sign the request
                var signer = new AWS4SignerForAuthorizationHeader()
                {
                    EndpointUri = uri,
                    HttpMethod = "GET",
                    Service = "sqs",
                    Region = region,
                };
                var authorization = signer.ComputeSignature(headers, requestParameters, AWS4SignerBase.EMPTY_BODY_SHA256,
                    config["AccessKey"], config["SecretKey"]);
                headers.Add("Authorization", authorization);

                string response = HttpHelpers.InvokeHttpRequest(uri, "GET", headers, null);
                //Console.WriteLine("Receipt for message deleted: " + receipt);

            }
        }
    }
}

File: RightslineDemoApplication_Java/RightslineDemoAppJavaSQS/src/main/java/com/rightsline/RestClient.java
package com.rightsline;

import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.rightsline.AWSSigner.AWS4SignerBase;
import com.rightsline.AWSSigner.AWS4SignerForAuthorizationHeader;
import com.rightsline.AWSSigner.HttpUtils;


import java.io.UnsupportedEncodingException;
import java.net.URL;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Timer;
import java.util.TimerTask;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class RestClient {
    private static String BaseUrl = "https://sqs.us-west-2.amazonaws.com/";
    private static HashMap<String, String> config = ConfigSetup.getConfigFile();
    private static int SecondsToPoll = 5;
    private static String individualMessageRegex = "\\{(.*?)\\}<";
    private static String receiptRegex = "ReceiptHandle>(.*?)<\\/Receipt";
    private static String entityRegex = "v2\\/(.*?)\\/";
    private static String MessagesToReceieve = "10";


    public static void DemoMethod(){
        String messages = getSQSMessages();
        ArrayList<JsonObject> jsonObjects = FilterXml(messages);

        ArrayList<String> receipts = FilterReceiptHandles(messages);
        DeleteMessage(receipts);
        Notify(jsonObjects, "catalog-item");
    }

    public static String getSQSMessages() {
        String region = config.get("Region");
        String endpointUri = BaseUrl + config.get("AccountId") + "/" + config.get("QueueName");
        String requestParameters = "Action=ReceiveMessage&MaxNumberOfMessages=" + MessagesToReceieve;
        HashMap<String, String> requestParams = new HashMap()
        {
            {
                put("Action", "ReceiveMessage");
                put("MaxNumberOfMessages", MessagesToReceieve);
            }
        };
        URL url = null;
        try {
            url = new URL(endpointUri + "?" + requestParameters);
        }
        catch(Exception e){
            e.printStackTrace();
        }
        HashMap headers = new HashMap<String, String>();
        headers.put(AWS4SignerBase.X_Amz_Content_SHA256, AWS4SignerBase.EMPTY_BODY_SHA256);
        AWS4SignerForAuthorizationHeader signer = new AWS4SignerForAuthorizationHeader(url, "GET", "sqs", region);
        String authorization = signer.computeSignature(headers, requestParams, AWS4SignerBase.EMPTY_BODY_SHA256, config.get("AccessKey"), config.get("SecretKey"));
        headers.put("Authorization", authorization);
        String response = HttpUtils.invokeHttpRequest(url, "GET", headers, null);
        return response;
    }
    public static void StartBackgroundMonitoring(){
        TimerTask PollTask = new TimerTask(){
            public void run(){
                String messages = getSQSMessages();
                ArrayList<JsonObject> jsonObjects = FilterXml(messages);
                Notify(jsonObjects, "catalog-item");
                Runnable deleteMessagesThread = () ->{
                    ArrayList<String> receipts = FilterReceiptHandles(messages);
                    DeleteMessage(receipts);
                };
                new Thread(deleteMessagesThread).start();
            }
        };
        Timer t = new Timer();
        t.schedule(PollTask, 2000, SecondsToPoll * 1000);
    }
    public static ArrayList<JsonObject> FilterXml(String XMLMessage){
        ArrayList<JsonObject> filteredMessages = new ArrayList<>();
        Pattern xmlFilter = Pattern.compile(individualMessageRegex);
        Matcher m = xmlFilter.matcher(XMLMessage);
        while(m.find()){
            //replace the &quot; with actual quotes from the XML message
            String match = m.group(0).replace("&quot;", "\"");
            filteredMessages.add(new JsonParser().parse(match.substring(0, match.length() - 1)).getAsJsonObject());
        }
        return filteredMessages;
    }
    public static ArrayList<String> FilterReceiptHandles(String XMLMessage){
        Pattern receiptFilter = Pattern.compile(receiptRegex);
        Matcher m = receiptFilter.matcher(XMLMessage);
        ArrayList<String> receipts = new ArrayList<>();
        while(m.find()){
//            try{

                receipts.add(m.group(1));
//            }
//            catch(Exception e){
//                e.printStackTrace();
//            }
        }
        return receipts;
    }
    private static void Notify(ArrayList<JsonObject> messages, String entitySearch){
        System.out.println("Poll at " + java.time.LocalDateTime.now());
        int numMessages = 0;
        Pattern entityPattern = Pattern.compile(entityRegex);
        for(JsonObject message : messages) {
            String entityUrl = message.get("entityUrl").toString();
            Matcher m = entityPattern.matcher(entityUrl);
            while(m.find()){
                String entityFound = m.group(1);
                if(entityFound.equals(entitySearch)){
                    numMessages++;
                    System.out.println("A(n) " + entityFound + " was " + message.get("action").toString() + ", URL:" + message.get("entityUrl"));
                }
            }
        }
        if(numMessages == 0){
            System.out.println("Recieved "+ messages.size() + " messages No messages regarding " + entitySearch + " were found.");
        }
    }

    /**
     *  Sends a request to Amazon to delete a batch of receipt handles for the messages received
      * @param receiptHandles
     */
    private static void DeleteMessage(ArrayList<String> receiptHandles){
        for(String receipt : receiptHandles){
            String region = config.get("Region");
            String endpointUri = BaseUrl + config.get("AccountId") + "/" + config.get("QueueName");
            String encodedReceipt = "";
            try {
                encodedReceipt = URLEncoder.encode(receipt, "UTF-8");
            } catch (UnsupportedEncodingException e) {
                e.printStackTrace();
            }
            //Receipt handle needs to be encoded for the URL but not in the requestParams hashmap
            String requestParameters = "Action=DeleteMessage&ReceiptHandle=" + encodedReceipt;
            HashMap<String, String> requestParams = new HashMap()
            {
                {
                    put("Action", "DeleteMessage");
                    put("ReceiptHandle", receipt);
                }
            };
            URL url = null;
            try {
                url = new URL(endpointUri + "?" + requestParameters);
            }
            catch(Exception e){
                e.printStackTrace();
            }
            HashMap headers = new HashMap<String, String>();
            headers.put(AWS4SignerBase.X_Amz_Content_SHA256, AWS4SignerBase.EMPTY_BODY_SHA256);
            AWS4SignerForAuthorizationHeader signer = new AWS4SignerForAuthorizationHeader(url, "GET", "sqs", region);
            String authorization = signer.computeSignature(headers, requestParams, AWS4SignerBase.EMPTY_BODY_SHA256, config.get("AccessKey"), config.get("SecretKey"));
            headers.put("Authorization", authorization);
            String response = HttpUtils.invokeHttpRequest(url, "GET", headers, null);
            //System.out.println(response);
        }
    }
}
